From 19dbf1d5d024dbfff19b2e07c5b1e1d24dee143e Mon Sep 17 00:00:00 2001
From: Yonit Halperin <yhalperi@redhat.com>
Date: Tue, 2 Feb 2010 16:05:44 +0200
Subject: [PATCH] spice: server: avoid video streaming of small images #558270

I added a lower limit to the size of images that are being streamed.
The limit is only active in "filter" video streaming mode.
This will prevent blurry animated icons.
---
 server/red_worker.c |   12 ++++++++++++
 1 files changed, 12 insertions(+), 0 deletions(-)

diff --git a/server/red_worker.c b/server/red_worker.c
index 015f184..8e5ca39 100644
--- a/server/red_worker.c
+++ b/server/red_worker.c
@@ -77,6 +77,7 @@
 #define RED_STREAM_FRAMES_START_CONDITION 20
 #define RED_STREAM_GRADUAL_FRAMES_START_CONDITION 0.2
 #define RED_STREAM_FRAMES_RESET_CONDITION 100
+#define RED_STREAM_MIN_SIZE (96 * 96)

 #define FPS_TEST_INTERVAL 1
 #define MAX_FPS 30
@@ -3311,6 +3312,17 @@ static inline void red_update_streamable(RedWorker *worker, Drawable *drawable,
         return;
     }

+    if (worker->streaming_video == STREAM_VIDEO_FILTER) {
+        Rect* rect;
+        int size;
+
+        rect = &drawable->qxl_drawable->u.copy.src_area;
+        size = (rect->right - rect->left) * (rect->bottom - rect->top);
+        if (size < RED_STREAM_MIN_SIZE) {
+            return;
+        }
+    }
+
     drawable->streamable = TRUE;
 }

-- 
1.6.6.1

